src/rust/engine/fs/src/directory.rs

```diff
@@ -0,0 +1,388 @@
+
+use crate::PathStat;
+
+lazy_static! {
+  pub static ref EMPTY_DIGEST_TREE: DigestTrie = DigestTrie(vec![].into());
+  pub static ref EMPTY_DIRECTORY_DIGEST: DirectoryDigest = DirectoryDigest {
+    digest: EMPTY_DIGEST,
+    tree: Some(EMPTY_DIGEST_TREE.clone()),
+  };
+}
+
+/// A Digest for a directory, optionally with its content stored as a DigestTrie.
+///
+/// If a DirectoryDigest has a DigestTrie reference, then its Digest _might not_ be persisted to
+/// the Store. If the DirectoryDigest does not hold a DigestTrie, then that Digest _must_ have been
+/// persisted to the Store (either locally or remotely). The field thus acts likes a cache in some
+/// cases, but in other cases is an indication that the tree must first be persisted (or loaded)
+/// before the Digest may be operated on.
+#[derive(Clone, DeepSizeOf)]
+pub struct DirectoryDigest {
+  pub digest: Digest,
+  pub tree: Option<DigestTrie>,
+}
+
+impl Eq for DirectoryDigest {}
+
+impl PartialEq for DirectoryDigest {
+  fn eq(&self, other: &Self) -> bool {
+    self.digest == other.digest
+  }
+}
+
+impl Debug for DirectoryDigest {
+  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    let tree = if self.tree.is_some() {
+      "Some(..)"
+    } else {
+      "None"
+    };
+    write!(f, "DirectoryDigest({:?}, tree: {})", self.digest, tree)
+  }
+}
+
+impl DirectoryDigest {
+  /// Creates a DirectoryDigest which asserts that the given Digest represents a Directory structure
+  /// which is persisted in a Store.
+  pub fn new(digest: Digest) -> Self {
+    Self { digest, tree: None }
+  }
+
+  /// Returns the digests reachable from this DirectoryDigest.
+  ///
+  /// If this DirectoryDigest has been persisted to disk (i.e., does not have a DigestTrie) then
+  /// this will only include the root.
+  pub fn digests(&self) -> Vec<Digest> {
+    if let Some(tree) = &self.tree {
+      let mut digests = tree.digests();
+      digests.push(self.digest);
+      digests
+    } else {
+      vec![self.digest]
+    }
+  }
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+struct Name(Intern<String>);
+known_deep_size!(0; Name);
+
+impl Deref for Name {
+  type Target = Intern<String>;
+
+  fn deref(&self) -> &Intern<String> {
+    &self.0
+  }
+}
+
+#[derive(DeepSizeOf)]
+pub enum Entry {
+  Directory(Directory),
+  File(File),
+}
+
+impl Entry {
+  fn name(&self) -> Name {
+    match self {
+      Entry::Directory(d) => d.name,
+      Entry::File(f) => f.name,
+    }
+  }
+}
+
+#[derive(DeepSizeOf)]
+pub struct Directory {
+  name: Name,
+  digest: Digest,
+  tree: DigestTrie,
+}
+
+impl Directory {
+  fn new(name: Name, entries: Vec<Entry>) -> Self {
+    Self::from_digest_tree(name, DigestTrie(entries.into()))
+  }
+
+  fn from_digest_tree(name: Name, tree: DigestTrie) -> Self {
+    Self {
+      name,
+      digest: tree.compute_root_digest(),
+      tree,
+    }
+  }
+
+  pub fn as_directory(&self) -> remexec::Directory {
```

Comment:
Could there be confusion for the reader from the struct being called `Directory` and this function being named `as_directory` (since it doesn't imply what the difference would be)?

Reply:
Great point. Maybe `as_remexec_directory`? In the age of autocomplete, I'm not worried about the extra chars.